### PR TITLE
Reader Teams: dedupe using lib/inflight instead

### PR DESCRIPTION
**Pull request by samouri**
For comparison with https://github.com/Automattic/wp-calypso/pull/11059.
This utilizes `lib/inflight` for request tracking instead of trying to introduce new middleware.

### CODE REVIEW

File: client/lib/inflight/index.js [hunk 1]
```diff
@@ -1,16 +1,24 @@
-var inflight = {};
+const inflight = new Set();
 
 // Utility methods to track inflight async requests
-module.exports = {
+export default {
```
Comment (blowery) [inline client/lib/inflight/index.js]: if you're going to go this road, I'd rather use named exports and fix up all the users of `lib/inflight` then export an object with props.

Comment (samouri) [inline client/lib/inflight/index.js]: sure

Comment (samouri) [inline client/lib/inflight/index.js]: @blowery: what about using named exports _and_ exporting an object with all the funcs.  that way its backwards compatible and we don't need to go mucking around with all the clients

**Comment by samouri** [comment]
Question for the reviewer:
whats the best way of testing the actual deduping within the data-layer?

File: client/lib/inflight/test/index.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ */
+import inflight from '../';
+
+const randomRequestKey = () => `${ Math.random() }`;
+const THE_ANSWER = 42;
+
+describe( 'inflight', ( ) => {
+	describe( 'promiseTracker', () => {
+		it( 'should mark an item as inflight while the request is ongoing', ( done ) => {
+			const key = randomRequestKey();
```
Comment (blowery) [inline client/lib/inflight/test/index.js]: why random? randomness in tests can lead to spurious failures....

File: client/lib/inflight/test/index.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+			const trackedPromise = inflight.promiseTracker( key, Promise.resolve( THE_ANSWER ) );
+
+			expect( inflight.requestInflight( key ) ).to.be.true;
+			trackedPromise.then( () => {
```
Comment (blowery) [inline client/lib/inflight/test/index.js]: you can just return the result of the `then` call.

File: client/lib/inflight/test/index.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+		it( 'should return the same data as the original promise', ( done ) => {
+			const key = randomRequestKey();
+			const trackedPromise = inflight.promiseTracker( key, Promise.resolve( THE_ANSWER ) );
+
+			trackedPromise.then( data => {
+				expect( data ).to.equal( THE_ANSWER );
+				done();
+			} );
+		} );
```
Comment (blowery) [inline client/lib/inflight/test/index.js]: need some tests for the error path too. You can use `Promise.reject` to trigger it.

**Comment by blowery** [comment]
> whats the best way of testing the actual deduping within the data-layer?

You need a test for the teams data layer bit that tries to push a duplicate request. You'll need to mock out `lib/wp` to wait on fulfilling the promise until after you've pushed a second request action. Then resolve it and validate that `wp.req.get` was only called once.

**Comment by samouri** [comment]
> You need a test for the teams data layer bit that tries to push a duplicate request. You'll need to mock out lib/wp to wait on fulfilling the promise until after you've pushed a second request action. Then resolve it and validate that wp.req.get was only called once.

@blowery I couldn't figure out a pretty way of doing that just yet :(.
First attempt is in https://github.com/Automattic/wp-calypso/pull/11069/files#diff-996b4418b115b23bf2ce3185e82e19beR83

File: client/state/data-layer/wpcom/read/teams/index.js [hunk 5]
```diff
@@ -3,9 +3,20 @@
  */
 import { READER_TEAMS_REQUEST, READER_TEAMS_RECEIVE } from 'state/action-types';
 import wpcom from 'lib/wp';
+import inflight from 'lib/inflight';
 
 export function handleTeamsRequest( store, action, next ) {
-	wpcom.req.get( '/read/teams', { apiVersion: '1.2' } )
+	const dedupeKey = READER_TEAMS_REQUEST;
+	if ( inflight.requestInflight( dedupeKey ) ) {
```
Comment (blowery) [inline client/state/data-layer/wpcom/read/teams/index.js]: we should also check if we've already loaded the teams and ignore the request if so. or maybe remember _when_ we loaded the teams and reload if it's been a bit.

Comment (samouri) [inline client/state/data-layer/wpcom/read/teams/index.js]: Did QueryComponents already do that?
I wouldn't want to introduce a new requirement

File: client/lib/inflight/test/index.js (hunk 2 above)
Comment (samouri) [inline client/lib/inflight/test/index.js]: removed randomness in https://github.com/Automattic/wp-calypso/pull/11069/commits/a6ea2c961b1f0c1aa6b4d4f0d59f79b209cb7307

File: client/state/data-layer/wpcom/read/teams/index.js (hunk 5 above)
Comment (blowery) [inline client/state/data-layer/wpcom/read/teams/index.js]: I don't think so, but if not, that was likely an oversight

Comment (blowery) [inline client/state/data-layer/wpcom/read/teams/index.js]: it can wait though

**Comment by blowery** [comment]
After further consideration, don't worry about testing deduping in unit tests just yet. Just make sure it works in the app; we're probably going to handle it elsewhere pretty soon.

**Comment by samouri** [comment]
I'm going to punt this for now because I think an `http-middleware` solution is Imminent

**Pull request closed without merging**